### PR TITLE
[SPARK-19631][CORE] OutputCommitCoordinator should not allow commits for already failed tasks

### PR DESCRIPTION
**Pull request by pwoody**
## What changes were proposed in this pull request?

Previously it was possible for there to be a race between a task failure and committing the output of a task. For example, the driver may mark a task attempt as failed due to an executor heartbeat timeout (possibly due to GC), but the task attempt actually ends up coordinating with the OutputCommitCoordinator once the executor recovers and committing its result. This will lead to any retry attempt failing because the task result has already been committed despite the original attempt failing.

This ensures that any previously failed task attempts cannot enter the commit protocol.

## How was this patch tested?

Added a unit test

### CODE REVIEW

**Comment by pwoody** [comment]
@JoshRosen @mccheah to review

File: core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala
Comment (ash211) [inline core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala]: can you put a comment here that the index into the `authorizedCommitters` array is the partitionId ?

**Review by ash211 — COMMENTED** [review]
Looks good @pwoody -- I can't imagine a case where we want the coordinator to authorize a task to commit that just came back as failed.

**Comment by vanzin** [comment]
ok to test

**Comment by ash211** [comment]
@vanzin are you right person to review this?

File: core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala]: Parameter indentation is incorrect. See "Indentation" section at http://spark.apache.org/contributing.html

File: core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala]: better:

```
stageState.failures.getOrElseUpdate(partition, mutable.Set()) += attemptNumber
```

File: core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala]: Same about parameter indentation.

File: core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala]: warning seems a little strong here; maybe info.

**Review by vanzin — COMMENTED** [review]
Can you expand the explanation of the race in the commit message? It's not clear how the race can happen with just what you wrote. The bug adds a little bit of info but it'd be better to have it properly explained here.

**Comment by pwoody** [comment]
Thanks for the feedback @vanzin

**Comment by vanzin** [comment]
Looks ok to me, but let me ping some others @squito @kayousterhout

File: core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala]: Why not define failures as an member variable (and initialize it there with an empty map), rather than forcing the caller to pass in an empty map?

**Review by kayousterhout — CHANGES_REQUESTED** [review]
I left one small comment inline, but other than that, this looks good.

@squito This commit makes me worried there are more bugs related to #16620.  For example, what if a task was OK'ed to commit, but then DAGScheduler decides to ignore it because of the epoch.  The DAGScheduler / TaskSetManager will attempt to re-run the task, but the output commit will never be OK'ed, which will cause the task to fail a bunch of times and the stage to get aborted.  Maybe this is OK because it's unlikely a stage will both be a shuffle map stage and also save output to HDFS?  Thoughts?

File: core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala]: ah sorry now that I see this I realized it probably makes sense to initialize `arr` in the StageState constructor too (so this line would look like `new StageState(maxPartitionId +1)`, and the `StageState` constructor just takes in `numPartitions`).  Would you mind making that change too?

Comment (pwoody) [inline core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala]: Yep sure thing, just pushed up the change.

**Review by kayousterhout — COMMENTED** [review]
LGTM assuming tests pass.  Let's wait to see if @squito has any comments here before merging.

File: core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala]: minor: the one place this is called, you've already looked up the state.  so you could take the `StageState` instead of the `StageId` (or even just inline the function entirely, though I think it may be a bit easier to understand as a helper method).

**Review by squito — APPROVED** [review]
lgtm

**Comment by squito** [comment]
@kayousterhout 
> This commit makes me worried there are more bugs related to #16620. For example, what if a task was OK'ed to commit, but then DAGScheduler decides to ignore it because of the epoch. The DAGScheduler / TaskSetManager will attempt to re-run the task, but the output commit will never be OK'ed, which will cause the task to fail a bunch of times and the stage to get aborted. Maybe this is OK because it's unlikely a stage will both be a shuffle map stage and also save output to HDFS? Thoughts?

yes, I think you are right, both about the bug, and that its pretty unlikely.  It looks like `SparkHadoopMapRedUtil` is a public class, so a user could write to hdfs inside a map stage, but that would be pretty weird.

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala]: I just realized there is still a potential race here -- what if `canCommit` happens, and then the Executor dies?  we don't know if the output has been written or not.  This change allows another task to commit its output.  That is good if the first task hadn't ever written its output.

But what if the earlier attempt had committed?  I *think* its OK for it to let another task commit its output over the original output.  (If it didn't, then we'd be back to the original scenario, with all future tasks failing b/c they couldn't commit their output.)

If that reasoning sounds correct, I think there should also be a test case where `canCommit()` comes before the task failure.

Comment (vanzin) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala]: > But what if the earlier attempt had committed?

I think there's a really tricky race here that I don't know if Spark is even able to fix. Basically:

- E1 starts to commit
- E1 loses connectivity with the driver, still committing
- E2 gets permission to commit from driver
- both E1 and E2 are committing their output files

Depending on how those output files are generated, you may either end up with corrupt output (i.e. output files from both executors) or task failures with partial output (E2 would fail to commit, you'd have some output from E1 in the final directory, task would be retried).

I believe that both tasks in this case would have the same set of output files, so one would overwrite the other; so the real problem is if E1 "wins" the race but leaves incomplete output around.

Comment (vanzin) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala]: BTW this is very unlikely to happen with filesystems that have atomic moves (or close enough to that), such as HDFS. If all attempts have the same set of outputs, you'd end up with the correct output eventually.

It might be trickier to reason about with filesystems that don't have that feature (hello S3), depending on the actual semantics of how (and if) tasks fail.

Comment (pwoody) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala]: Yeah - this should be fine and is how it currently works as well.

If a failure comes in after the authorization, then the task may or may not commit before failure. The coordinator will release the authorization once it gets the failure, and the next task attempt will check and possibly delete any existing data left over while attempting its own commit.

Happy to add the test though.

Comment (pwoody) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala]: sorry that was in response to @squito 

@vanzin Task attempt 1 could very well remove task attempt 2s output even after the second attempt has reported success back to the driver and then get forcibly killed during its own commit.

Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala]: so, while we are fixing a problem that spark fails, when it should succeed -- aren't we creating another problem, that now spark says it succeeds, when really it has failed (it may have written corrupt data)?  That seems like a worse situation.

Maybe this is fine with hdfs and an atomic move -- but wouldn't this be a big problem on S3?

Comment (pwoody) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala]: The race described above doesn't occur because of this change, it has always existed as far as I know. This change blocks anyone from doing `canCommit` if they've already failed (which causes retries to infinitely loop when the commit is done).

Comment (ash211) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala]: My understanding is that this change doesn't affect any task that's mid-commit, it only changes the pre-commit portion of a task attempt's lifecycle to prevent failed tasks (=> heartbeat timeouts as reported by the DAGScheduler) from receiving authorization to commit.  Is there a new way which two task attempts can be authorized simultaneously that didn't exist before?

Comment (vanzin) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala]: Yes, that seems accurate. In my example above, this change wouldn't change anything, since it's solving a different race (driver would not allow E1 to commit if it thinks it's dead).

**Comment by vanzin** [comment]
retest this please

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala]: sorry if I am being really dense, but it still seems to me like in this particular scenario, we're taking broken behavior, fixing it in some cases, and making it worse in others.  

suppose E1 got permission to commit, then lost connectivity to the driver (or missed heartbeats etc), but continued to try to commit.  then E2 asks to commit.

Before, we might have ended up with an infinite loop, where E1 never finishes committing, and E2 never gets to commit.  Similarly, all future attempts don't get to commit, but we don't even fail the task set because taskcommitdenied doesn't count towards failing a taskset, so it just keeps retrying.

After this change, E2 gets to commit immediately after E1 loses connectivity to the driver.  E1 may or may not commit at any time.  If E1 doesn't commit, great.  If E1 does commit, then in most scenarios, things will still be fine.  But sometimes, the two commits will stomp on each other. 

so we've narrowed the scenarios with incorrect behavior -- but the behavior has gone from an infinite loop (bad), to jobs appearing to succeed when they have actually written corrupt data (worse, IMO).

Comment (pwoody) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala]: @squito This PR does not affect anything after E1 gets permission to commit, the race you describe is definitely possible and has existed before. This change here only makes it such that if E1 fails before asking to commit, then it is blacklisted from being authorized to commit.

Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala]: ok, I finally get it.  I was thinking this change was doing something different.  Sorry it took me a while.

That said, I realize there *is* another issue here.  I tried to update the test to confirm that no other task would commit once there was an executor failure, like so:

```scala
  test("SPARK-19631: Do not allow failed attempts to be authorized for committing") {
    val stage: Int = 1
    val partition: Int = 1
    val failedAttempt: Int = 0
    outputCommitCoordinator.stageStart(stage, maxPartitionId = 1)
    outputCommitCoordinator.taskCompleted(stage, partition, attemptNumber = failedAttempt,
      reason = ExecutorLostFailure("0", exitCausedByApp = true, None))
    // if we get a request to commit after we learn the executor failed, we don't authorize
    // the task to commit, so another attempt can commit.
    assert(!outputCommitCoordinator.canCommit(stage, partition, failedAttempt))
    assert(outputCommitCoordinator.canCommit(stage, partition, failedAttempt + 1))
    // but if we get an executor failure *after* we authorize a task to commit, we never let
    // another task commit.  Unfortunately, we just don't know what the status is of the first task,
    // so we can't safely let any other task proceed.
    outputCommitCoordinator.taskCompleted(stage, partition, attemptNumber = failedAttempt + 1,
      reason = ExecutorLostFailure("0", exitCausedByApp = true, None))
    assert(!outputCommitCoordinator.canCommit(stage, partition, failedAttempt + 2))
  }
```

This test fails at the final assert, because the executor failure *does* [clear the authorized commiter](https://github.com/apache/spark/blob/8aa560b75e6b083b2a890c52301414285ba35c3d/core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala#L143).  But this PR doesn't change that at all -- an equivalent check in master would also fail.

**Comment by squito** [comment]
lgtm, sorry for the noise

File: core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/OutputCommitCoordinatorSuite.scala]: https://issues.apache.org/jira/browse/SPARK-19790

File: core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/OutputCommitCoordinator.scala]: can you fix the indentation here? (should be +2 spaces)

**Comment by ash211** [comment]
Any last changes before merging?

**Comment by kayousterhout** [comment]
LGTM -- I merged this into master

**Pull request closed without merging**
